Userland/Libraries/LibWeb/CSS/StyleValues/EasingStyleValue.cpp

```diff
@@ -57,30 +64,154 @@ bool EasingStyleValue::CubicBezier::operator==(Web::CSS::EasingStyleValue::Cubic
     return x1 == other.x1 && y1 == other.y1 && x2 == other.x2 && y2 == other.y2;
 }
 
-double EasingStyleValue::Linear::evaluate_at(double input_progress, bool) const
+// https://drafts.csswg.org/css-easing/#linear-canonicalization
+EasingStyleValue::Linear::Linear(Vector<EasingStyleValue::Linear::Stop> stops)
 {
-    return input_progress;
+    // To canonicalize a linear() function’s control points, perform the following:
+
+    // 1. If the first control point lacks an input progress value, set its input progress value to 0.
+    if (!stops.first().input.has_value())
+        stops.first().input = 0;
+
+    // 2. If the last control point lacks an input progress value, set its input progress value to 1.
+    if (!stops.last().input.has_value())
+        stops.last().input = 1;
+
+    // 3. If any control point has an input progress value that is less than
+    // the input progress value of any preceding control point,
+    // set its input progress value to the largest input progress value of any preceding control point.
+    double largest_input = 0;
+    for (auto stop : stops) {
+        if (stop.input.has_value()) {
+            if (stop.input.value() < largest_input) {
+                stop.input = largest_input;
+            } else {
+                largest_input = stop.input.value();
+            }
+        }
+    }
+
+    // 4. If any control point still lacks an input progress value,
+    // then for each contiguous run of such control points,
+    // set their input progress values so that they are evenly spaced
+    // between the preceding and following control points with input progress values.
+    Optional<size_t> run_start_idx;
+    for (size_t idx = 0; idx < stops.size(); idx++) {
+        auto stop = stops[idx];
+        if (stop.input.has_value() && run_start_idx.has_value()) {
+            // this stop is immediately after a run
+            // set inputs of [start, idx-1] stops to be evenly spaced between start-1 and idx
```

Comment:
To make it clear what's spec and what isn't, please prefix these with `Note:` or `NOTE:`. (I think we use both which is maybe not ideal - try whichever one the spec itself doesn't use. 😅)

Reply:
Done :+1: